arq/parser.py

```diff
@@ -0,0 +1,126 @@
+#  type: ignore
```

Comment:
This is my main question about this PR (and about migrating to aioredis v2 in general).

In general, it's a real shame that the developers of aioredis 2 decided to constraint the encode/decode decision to be made on the connection/pool, not on an individual command. I assume this was so they code do (or not do) the encoding in hiredis and improve performance?

What you've written here is very cool and a sophisticated approach, but it doesn't actually avoid backwards compatibility problems, and in some ways makes the logic within the library more complex. I also wonder if it will incur a performance penalty?

It makes me sad to say it, but I wonder if we should removing this and just use `.decode()` everywhere that we want a string?

Always returning bytes and forcing the user/developer to call `.decode()` is definitely the explicit keep-it-simple-stupid approach.

For the tests we can just compare to bytes, no need for either `with self.pool.encoder_context...` or `.decode()`.

It will mean in application logic that users (including me!) have to add `.decode()` in a lot of places, but the alternative is to add `with self.pool.encoder_context...` in nearly as many places - there's no route that doesn't cause a big fat change to the API.

Reply:
I think their reasoning was just to port redis-py. Because redis-py does not support decode on per-request basis, they also decided not to add it. But redis-py maintainer actually thought of adding this feature, they even add to hiredis an ability to change encoding settings at runtime https://github.com/redis/hiredis-py/pull/96 (and we use it in `ContextAwareHiredisParser`). Also I am pretty sure that I saw a branch in `redis-py` repository where they started working on the encoding context, but it seems it was removed at some point (so may be they decided against it).

In terms of performance, as I understand, changing encoding on HiRedis should not be a huge overhead in terms of performance, this method just updates parameters. But I am not sure about how much context variables handling can influence performance. I can run some tests later.

But if you really think that we should stick with simpler approcah, I can remove context encoding and return to `decode` calls (basically what I did with first set of changes).

Reply:
They even mention encoding context manager in `lock.py` https://github.com/aio-libs/aioredis-py/blob/master/aioredis/lock.py#L235 (I guess it was just copied from `redis-py`).